pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1916,20 +1916,20 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
         if (messageIds == null || messageIds.isEmpty()) {
             return CompletableFuture.completedFuture(Collections.emptyList());
         }
-        List<MessageIdData> data = new ArrayList<>(messageIds.size());
```

Comment:
Couldn't we fix this by using `CopyOnWriteArrayList` or wrapping with `Collections.synchronizedList`?

Reply:
(since the futures based solution is fine, this comment can be ignored)